### PR TITLE
[Flax] Fix incomplete batches in example scripts

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

Currently in our Flax examples scripts, we drop the last incomplete batch during training and inference:
https://github.com/huggingface/transformers/blob/09178705101b9803e7b9ea7f79a46b4c242dd4bf/examples/flax/summarization/run_summarization_flax.py#L350
We do this for two reasons:

1. Because XLA is not shape polymorphic, forming the last batch with shape different from the preceding batches triggers a recompilation of the `pmap`'d function .
2. If the batch size is not divisible by the number devices, then the last step must be executed on a single device (or a subset of devices), potentially leading to OOMs.

During training, dropping the last batch isn't an issue: since we shuffle the data and train for multiple epochs, all of the training data is eventually used and the effects of dropping the last batch amortised.

However, during evaluation and prediction, dropping the last batch leads to incorrect results: since we don't account for the examples in the last batch, we do not evaluate over the whole dataset, and thus have partial results.

This PR corrects for the incomplete batches in the relevant Flax training examples.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/flax/summarization/run_summarization_flax.py [hunk 1]
```diff
@@ -847,7 +846,7 @@ def generate_step(params, batch):
         # train
         for _ in tqdm(range(steps_per_epoch), desc="Training...", position=1, leave=False):
             batch = next(train_loader)
-            state, train_metric = p_train_step(state, batch)
+            state, train_metric = pad_shard_unpad(p_train_step)(state, batch)
```
Comment (patil-suraj) [inline examples/flax/summarization/run_summarization_flax.py:849]: This is really nit!

File: examples/flax/summarization/run_summarization_flax.py [hunk 2]
```diff
@@ -337,24 +338,22 @@ def replicate(self):
...
 
-    batch_idx = batch_idx[: steps_per_epoch * batch_size]  # Skip incomplete batch.
-    batch_idx = batch_idx.reshape((steps_per_epoch, batch_size))
+    batch_idx = np.array_split(batch_idx, steps_per_epoch)
 
     for idx in batch_idx:
         batch = dataset[idx]
-        batch = {k: jnp.array(v) for k, v in batch.items()}
-
-        batch = shard(batch)
```
Comment (patil-suraj) [inline examples/flax/summarization/run_summarization_flax.py:357]: does `pad_shard_unpad` also takes care of sharding the batch ? (seems like it, but just to confirm)

**Review by patil-suraj — APPROVED** [review]
Thanks, this look good to me! The `pad_shard_unpad` trick is neat!

File: examples/flax/summarization/run_summarization_flax.py (hunk 2 above)
Comment (sanchit-gandhi) [reply examples/flax/summarization/run_summarization_flax.py:357]: Yes sir!

**Review by patrickvonplaten — APPROVED** [review]
Thanks for fixing @sanchit-gandhi !

**Comment by sanchit-gandhi** [comment]
Just waiting to double check that the slow tests pass from [`test_flax_examples.py`](https://github.com/huggingface/transformers/blob/main/examples/flax/test_flax_examples.py) before merging. Working with @patil-suraj to verify this ✅

**Comment by patrickvonplaten** [comment]
@patil-suraj @sanchit-gandhi can we merge this one?

**Comment by sanchit-gandhi** [comment]
Just verifying the slow tests from [`test_flax_examples.py`](https://github.com/huggingface/transformers/blob/main/examples/flax/test_flax_examples.py) pass on a v3-8!
